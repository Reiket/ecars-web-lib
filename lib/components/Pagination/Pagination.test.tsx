import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';
import {
  Pagination,
  PAGINATION_ARROW_LEFT_TEST_ID,
  PAGINATION_ARROW_RIGHT_TEST_ID,
  PAGINATION_TEST_ID,
} from '@/components/Pagination/constants';

const props = {
  pages: [1, 2, 3],
  currentPage: 1,
  onPageClick: jest.fn(),
  onNextClick: jest.fn(),
  onPrevClick: jest.fn(),
};

describe('Pagination component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders component with all page buttons and arrows', () => {
    const {container} = render(<Pagination {...props} />);

    const paginationElement = screen.getByTestId(PAGINATION_TEST_ID);
    const leftArrowElement = screen.getByTestId(PAGINATION_ARROW_LEFT_TEST_ID);
    const rightArrowElement = screen.getByTestId(PAGINATION_ARROW_RIGHT_TEST_ID);
    expect(paginationElement).toBeInTheDocument();
    expect(leftArrowElement).toBeInTheDocument();
    expect(rightArrowElement).toBeInTheDocument();
    props.pages.forEach((page) => {
      expect(screen.getByText(String(page))).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });

  test('disables prev arrow on first page', () => {
    const firstPage = props.pages.at(0) ?? 0;
    const lastPage = props.pages.at(-1) ?? 0;
    const {container, rerender} = render(
      <Pagination
        {...props}
        currentPage={firstPage}
      />,
    );

    const leftArrowElement = screen.getByTestId(PAGINATION_ARROW_LEFT_TEST_ID).parentElement;
    const rightArrowElement = screen.getByTestId(PAGINATION_ARROW_RIGHT_TEST_ID).parentElement;

    expect(leftArrowElement).toBeDisabled();

    rerender(
      <Pagination
        {...props}
        currentPage={lastPage}
      />,
    );

    expect(rightArrowElement).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  test('calls onPageClick when page button is clicked', () => {
    render(
      <Pagination
        {...props}
        currentPage={2}
      />,
    );

    fireEvent.click(screen.getByText('2'));
    expect(props.onPageClick).toHaveBeenCalledWith(2);
  });
  test('marks the current page as selected', () => {
    render(<Pagination {...props} />);

    expect(screen.getByText('1')).toHaveClass('_current-page');
  });
  test('calls onPrevClick and onNextClick', () => {
    render(
      <Pagination
        {...props}
        currentPage={2}
      />,
    );

    const leftArrowElement = screen.getByTestId(PAGINATION_ARROW_LEFT_TEST_ID).parentElement;
    const rightArrowElement = screen.getByTestId(PAGINATION_ARROW_RIGHT_TEST_ID).parentElement;

    if (leftArrowElement) {
      fireEvent.click(leftArrowElement);
      expect(props.onPrevClick).toHaveBeenCalled();
    }

    if (rightArrowElement) {
      fireEvent.click(rightArrowElement);
      expect(props.onNextClick).toHaveBeenCalled();
    }
  });
  test('disables all buttons and arrows when loading', () => {
    render(
      <Pagination
        {...props}
        isLoading
      />,
    );

    const leftArrow = screen.getByTestId(PAGINATION_ARROW_LEFT_TEST_ID).parentElement;
    const rightArrow = screen.getByTestId(PAGINATION_ARROW_RIGHT_TEST_ID).parentElement;
    expect(leftArrow).toBeDisabled();
    expect(rightArrow).toBeDisabled();

    props.pages.forEach((page) => {
      expect(screen.getByText(String(page))).toBeDisabled();
    });

    fireEvent.click(screen.getByText('1'));
    if (leftArrow) {
      fireEvent.click(leftArrow);
    }
    if (rightArrow) {
      fireEvent.click(rightArrow);
    }
    expect(props.onPageClick).not.toHaveBeenCalled();
    expect(props.onPrevClick).not.toHaveBeenCalled();
    expect(props.onNextClick).not.toHaveBeenCalled();
  });
});
