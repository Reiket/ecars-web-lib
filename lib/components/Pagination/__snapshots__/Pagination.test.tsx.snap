// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination component > disables prev arrow on first page 1`] = `
<div>
  <div
    class="pagination"
    data-testid="paginationTestId"
  >
    <div
      class="pagination__body"
    >
      <button
        class="pagination__button"
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowLeftTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </button>
      <button
        class="pagination__button"
      >
        1
      </button>
      <button
        class="pagination__button"
      >
        2
      </button>
      <button
        class="pagination__button _current-page"
      >
        3
      </button>
      <button
        class="pagination__button"
        disabled=""
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowRightTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination component > renders component with all page buttons and arrows 1`] = `
<div>
  <div
    class="pagination"
    data-testid="paginationTestId"
  >
    <div
      class="pagination__body"
    >
      <button
        class="pagination__button"
        disabled=""
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowLeftTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </button>
      <button
        class="pagination__button _current-page"
      >
        1
      </button>
      <button
        class="pagination__button"
      >
        2
      </button>
      <button
        class="pagination__button"
      >
        3
      </button>
      <button
        class="pagination__button"
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowRightTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination component disables prev arrow on first page 1`] = `
<div>
  <div
    class="pagination"
    data-testid="paginationTestId"
  >
    <div
      class="pagination__body"
    >
      <button
        class="pagination__button"
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowLeftTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </button>
      <button
        class="pagination__button"
      >
        1
      </button>
      <button
        class="pagination__button"
      >
        2
      </button>
      <button
        class="pagination__button _current-page"
      >
        3
      </button>
      <button
        class="pagination__button"
        disabled=""
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowRightTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Pagination component renders component with all page buttons and arrows 1`] = `
<div>
  <div
    class="pagination"
    data-testid="paginationTestId"
  >
    <div
      class="pagination__body"
    >
      <button
        class="pagination__button"
        disabled=""
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowLeftTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M15.41 16.59 10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
          />
        </svg>
      </button>
      <button
        class="pagination__button _current-page"
      >
        1
      </button>
      <button
        class="pagination__button"
      >
        2
      </button>
      <button
        class="pagination__button"
      >
        3
      </button>
      <button
        class="pagination__button"
        type="button"
      >
        <svg
          class="pagination__arrow"
          data-testid="paginationArrowRightTestId"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
